homeassistant/components/xiaomi_miio/sensor.py

```diff
@@ -135,17 +155,234 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
     if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
         host = config_entry.data[CONF_HOST]
         token = config_entry.data[CONF_TOKEN]
+        model = config_entry.data[CONF_MODEL]
         name = config_entry.title
         unique_id = config_entry.unique_id
 
         _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
 
-        device = AirQualityMonitor(host, token)
-        entities.append(XiaomiAirQualityMonitor(name, device, config_entry, unique_id))
+        if model == MODEL_AIRPURIFIER_3C:
+            air_purifier = AirPurifierMB4(host, token)
+            entities.append(
```

Comment:
The problem with this is that we are now creating 3 sensor entities + 1 fan entity that all have seperate polling updates (`async def async_update(self):`). While a single update contains all information for all 4 entities. Besides these updates are not synced, so they may happen all at once.
I am not sure what will happen, but this might cause stability problems, not sure how well xiaomi devices handel a lot of requests at once. (Besides the fact that this is a waste of network resources and resources of the device running Hass).

Therefore the preferred way to go would be using a DataUpdateCoordinator.
Unfortunateley that is not so trivial (see https://github.com/rytilahti/python-miio/issues/929#issuecomment-787442355 and https://github.com/home-assistant/core/pull/46964#discussion_r587736441)

So for now I think we can except this duplicate updates, but this defenately needs to be changed in the future.
@rytilahti what do you think?

Reply:
If I understand correctly, DataUpdateCoordinator should be implemented globally (in the XiaomiMiioEntity class), I guess this may break other devices, so I think it's an argument for another issue / PR.
Anyway, I'm preparing a commit to fix the other issues.